### PR TITLE
Add linkerd viz authz command

### PR DESCRIPTION
**Pull request by adleong**
Similarly to the `linkerd authz` command which lists all authorizations for a given resource and `linkerd viz stat` which can show metrics for policy resources, we introduce a `linkerd viz authz` command which shows metrics for server authorizations broken down by server for a given resource.  It also shows the rate of unauthorized requests to each server.  This is helpful for seeing a breakdown of which authorizations are being used and what proportion of traffic is being rejected.  For example:

```console
> linkerd viz authz -n emojivoto deploy
SERVER       AUTHZ            SUCCESS     RPS  LATENCY_P50  LATENCY_P95  LATENCY_P99  
emoji-grpc   emoji-grpc       100.00%  1.8rps          1ms          1ms          1ms  
prom         prom-prometheus        -       -            -            -            -  
voting-grpc  [UNAUTHORIZED]         -  0.9rps            -            -            -  
web-http     web-public        50.00%  1.8rps          4ms        190ms        198ms
```

This shows us a few things right away:

* all traffic to the emoji-grpc server is authorized by the emoji-grpc server authorization
* the prom server defines a prom-prometheus server authorization, but it is not receiving any traffic
* the voting-grpc server has no server authorizations, and thus all 0.9rps is getting rejected

TODO:
* this command requires the CLI to make many API calls.  it could be sped up by making some of these calls in parallel or by moving more logic into the metrics-api controller to take better advantage of the metrcs-api client-go caches
* add support for json output
* improve CLI help output

### CODE REVIEW

**Comment by adleong** [comment]
I am planning to rebase once https://github.com/linkerd/linkerd2/pull/6785 has been merged, which should clean up the diff.

File: pkg/k8s/policy.go [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+
+// ServerAndAuthorization holds a server name and serverauthorization name.
+type ServerAndAuthorization struct {
+	Server              string
+	ServerAuthorization string
+}
+
+type id struct{ name, namespace string }
+
+var sazGVR = schema.GroupVersionResource{
```
Comment (Pothulapati) [inline pkg/k8s/policy.go:24]: Looks like we can use these and remove them from https://github.com/linkerd/linkerd2/blob/alex/viz-authz/viz/metrics-api/stat_summary.go#L28

File: cli/cmd/metrics_test.go [hunk 2]
```diff
@@ -116,7 +116,7 @@ spec:
 	// However, only pod-1 is owned by deploy-1 according to the owner references.
 	// Owner references should be considered authoritative to resolve ambiguity
 	// when deployments have overlapping seletors.
-	pods, err := getPodsFor(context.Background(), k8sClient, "ns", "deploy/deploy-1")
+	pods, err := k8s.GetPodsFor(context.Background(), k8sClient, "ns", "deploy/deploy-1")
```
Comment (Pothulapati) [inline cli/cmd/metrics_test.go:119]: As this exclusively tests the `GetPodsFor`. Should we move it into the k8s tests.

File: viz/metrics-api/prometheus.go [hunk 3]
```diff
@@ -118,8 +118,14 @@ func promDstGroupByLabelNames(resource *pb.Resource) model.LabelNames {
 func promQueryLabels(resource *pb.Resource) model.LabelSet {
 	set := model.LabelSet{}
 	if resource != nil {
-		if resource.Name != "" && resource.GetType() != k8s.Service {
-			set[promResourceType(resource)] = model.LabelValue(resource.Name)
+		if resource.Name != "" {
+			if resource.GetType() == k8s.Server {
```
Comment (Pothulapati) [inline viz/metrics-api/prometheus.go:122]: Is this change required? In `stat_summary.go`, We use the resource labels directly? We can keep this change and remove the manual setting there? :thinking:

**Review by Pothulapati — COMMENTED** [review]
This looks great, and works as expected. Really liked how these commands are consistent, and shows extra metrics with the `viz` prefix command.

```bash

on ⛵ kind-kind  linkerd2 on 🌱 alex [📦📝🤷‍] via 🐼 v1.16.7 via  via ⚙️ v1.55.0took 2s
➜ ./bin/go-run cli viz authz deploy/emoji -n emojivoto                                               ~/work/linkerd2
SERVER      AUTHZ            SUCCESS     RPS  LATENCY_P50  LATENCY_P95  LATENCY_P99
emoji-grpc  emoji-grpc       100.00%  1.5rps          0ms          0ms          0ms
emoji-grpc  [UNAUTHORIZED]         -  0.3rps            -            -            -
prom        emoji-grpc             -       -            -            -            -
prom        prom-prometheus        -       -            -            -            -
```

> this command requires the CLI to make many API calls. it could be sped up by making some of these calls in parallel or by moving more logic into the metrics-api controller to take better advantage of the metrcs-api client-go caches

Yes, It does indeed perform a good number of k8s queries as we need to find the relationships. I'm not sure moving this logic to the `metrics-api` helps as we don't do any of the informer stuff for the policy resources. But we should get some benefits in going from the workload -> pods with the caching stuff.

I agree with the todo's mentioned, and I'm okay with trying the perf improvements in a separate PR.

File: viz/metrics-api/prometheus.go (hunk 3 above)
Comment (adleong) [reply viz/metrics-api/prometheus.go:122]: This was to fix a bug when calling stat summary with a server or saz as the ToResource.  In this case it wasn't rewriting the label name and was attempting to query prometheus with `dst_server=` instead of `srv=` or something like that.

**Review by Pothulapati — APPROVED** [review]
LGTM, except for the last two todo's i.e json output, and help info

File: viz/cmd/authz.go [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+			if options.namespace == "" {
+				options.namespace = pkgcmd.GetDefaultNamespace(kubeconfigPath, kubeContext)
+			}
+
+			k8sAPI, err := k8s.NewAPI(kubeconfigPath, kubeContext, impersonate, impersonateGroup, 0)
+			if err != nil {
+				return err
+			}
+			// The gRPC client is concurrency-safe, so we can reuse it in all the following goroutines
+			// https://github.com/grpc/grpc-go/issues/682
```
Comment (alpeb) [inline viz/cmd/authz.go:55]: Not applicable in this case
```suggestion
```

**Review by alpeb — COMMENTED** [review]
The `--selector` flag is being applied to srv and saz resources; shouldn't it instead apply to the main resource being queried?
Also as a further improvement down the road, it'd be nice for this command and `linkerd authz` to show what resource each line applies to, when only specifying the workload type (like in your example above). Perhaps showing the pod selector in `-owide` should do :wink:

**Review by alpeb — APPROVED** [review]
Thanks! :+1:
